File: app/src/main/java/com/github/gotify/api/Callback.java
package com.github.gotify.api;

import android.app.Activity;
import com.github.gotify.log.Log;
import retrofit2.Call;
import retrofit2.Response;

public class Callback<T> {
    private final SuccessCallback<T> onSuccess;
    private final ErrorCallback onError;

    private Callback(SuccessCallback<T> onSuccess, ErrorCallback onError) {
        this.onSuccess = onSuccess;
        this.onError = onError;
    }

    public static <T> retrofit2.Callback<T> callInUI(
            Activity context, SuccessCallback<T> onSuccess, ErrorCallback onError) {
        return call(
                (data) -> context.runOnUiThread(() -> onSuccess.onSuccess(data)),
                (e) -> context.runOnUiThread(() -> onError.onError(e)));
    }

    public static <T> retrofit2.Callback<T> call() {
        return call((e) -> {}, (e) -> {});
    }

    public static <T> retrofit2.Callback<T> call(
            SuccessCallback<T> onSuccess, ErrorCallback onError) {
        return new RetrofitCallback<>(merge(of(onSuccess, onError), errorCallback()));
    }

    private static <T> Callback<T> of(SuccessCallback<T> onSuccess, ErrorCallback onError) {
        return new Callback<>(onSuccess, onError);
    }

    private static <T> Callback<T> errorCallback() {
        return new Callback<>((ignored) -> {}, (error) -> Log.e("Error while api call", error));
    }

    private static <T> Callback<T> merge(Callback<T> left, Callback<T> right) {
        return new Callback<>(
                (data) -> {
                    left.onSuccess.onSuccess(data);
                    right.onSuccess.onSuccess(data);
                },
                (error) -> {
                    left.onError.onError(error);
                    right.onError.onError(error);
                });
    }

    public interface SuccessCallback<T> {
        void onSuccess(T data);
    }

    public interface ErrorCallback {
        void onError(ApiException t);
    }

    private static final class RetrofitCallback<T> implements retrofit2.Callback<T> {

        private Callback<T> callback;

        private RetrofitCallback(Callback<T> callback) {
            this.callback = callback;
        }

        @Override
        public void onResponse(Call<T> call, Response<T> response) {
            if (response.isSuccessful()) {
                callback.onSuccess.onSuccess(response.body());
            } else {
                callback.onError.onError(new ApiException(response));
            }
        }

        @Override
        public void onFailure(Call<T> call, Throwable t) {
            callback.onError.onError(new ApiException(t));
        }
    }
}

File: app/src/main/kotlin/com/github/gotify/api/Callback.kt
package com.github.gotify.api

import android.app.Activity
import com.github.gotify.api.Callback.ErrorCallback
import com.github.gotify.api.Callback.SuccessCallback
import org.tinylog.kotlin.Logger
import retrofit2.Call
import retrofit2.Response

internal class Callback<T> private constructor(
    private val onSuccess: SuccessCallback<T>,
    private val onError: ErrorCallback
) {
    fun interface SuccessCallback<T> {
        fun onSuccess(response: Response<T>)
    }

    fun interface SuccessBody<T> : SuccessCallback<T> {
        override fun onSuccess(response: Response<T>) {
            onResultSuccess(response.body() ?: throw ApiException("null response", response))
        }

        fun onResultSuccess(data: T)
    }

    fun interface ErrorCallback {
        fun onError(t: ApiException)
    }

    private class RetrofitCallback<T>(private val callback: Callback<T>) : retrofit2.Callback<T> {
        override fun onResponse(call: Call<T>, response: Response<T>) {
            if (response.isSuccessful) {
                callback.onSuccess.onSuccess(response)
            } else {
                callback.onError.onError(ApiException(response))
            }
        }

        override fun onFailure(call: Call<T>, t: Throwable) {
            callback.onError.onError(ApiException(t))
        }
    }

    companion object {
        fun <T> callInUI(
            context: Activity,
            onSuccess: SuccessCallback<T>,
            onError: ErrorCallback
        ): retrofit2.Callback<T> {
            return call(
                onSuccess = { response -> context.runOnUiThread { onSuccess.onSuccess(response) } },
                onError = { exception -> context.runOnUiThread { onError.onError(exception) } }
            )
        }

        fun <T> call(
            onSuccess: SuccessCallback<T> = SuccessCallback {},
            onError: ErrorCallback = ErrorCallback {}
        ): retrofit2.Callback<T> {
            return RetrofitCallback(merge(of(onSuccess, onError), errorCallback()))
        }

        private fun <T> of(onSuccess: SuccessCallback<T>, onError: ErrorCallback): Callback<T> {
            return Callback(onSuccess, onError)
        }

        private fun <T> errorCallback(): Callback<T> {
            return Callback(
                onSuccess = {},
                onError = { exception -> Logger.error(exception, "Error while api call") }
            )
        }

        private fun <T> merge(left: Callback<T>, right: Callback<T>): Callback<T> {
            return Callback(
                onSuccess = { data ->
                    left.onSuccess.onSuccess(data)
                    right.onSuccess.onSuccess(data)
                },
                onError = { exception ->
                    left.onError.onError(exception)
                    right.onError.onError(exception)
                }
            )
        }
    }
}
